### PR TITLE
NO_MERGE | Feature/stock_allocation_mode

### PR DESCRIPTION
**Pull request by ViljoenFritz**
Summary of this change:
Add new column 'stock_allocation_mode' (text) to loads

CHANGELOG entry:

### Added
- [Add exclude to BaseRepo select_values](https://github.com/NoSoft-SA/nspack/pull/464/commits/d152f033fe4ea78bd786519462dbb14bcb51ff64)
### Changed
- [Add client_fg_rules stock_allocation_modes](https://github.com/NoSoft-SA/nspack/pull/464/commits/b879938637d68b633d8319a951e7fcc0baf850be)
- [Add load_stock_allocation_modes](https://github.com/NoSoft-SA/nspack/pull/464/commits/c234359fc8b9357cefaf8c580ab38601ec4612d5)

How to test this code via the UI:

### CODE REVIEW

File: lib/base_repo.rb [hunk 1]
```diff
@@ -342,9 +344,14 @@ def select_values_in_order(table_name, columns, order:, where: {}, descending: f
   # @param columns [Symbol,Array] the column (or array of columns) to query.
   # @param args [Hash] the where-clause conditions. Optional.
   # @return [Array] the values from the column(s) of each row.
-  def select_values(table_name, columns, args = nil)
+  def select_values(table_name, columns, args = {})
```
Comment (jcsjcs) [inline lib/base_repo.rb:347]: This is fine, just modify the yarddoc to explain how args can include `where` or `exclude` hashes.

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: lib/client_rules/client_fg_rules.rb [hunk 2]
```diff
@@ -437,16 +427,16 @@ def local_pallets_to_stock_via_pi?(explain: false)
       setting(:set_local_stock_via_pi)
     end
 
-    def enable_depot_returns?(explain: false)
-      return 'Allow allocation and shipping of failed govt_inspection pallets (must be inspected pallets).' if explain
+    def stock_allocation_modes(explain: false, mode: nil)
+      valid_modes = { standard: 'Stock Pallets',
```
Comment (jcsjcs) [inline lib/client_rules/client_fg_rules.rb:431]: These should be in AppConst.

File: lib/client_rules/client_fg_rules.rb [hunk 3]
```diff
@@ -437,16 +427,16 @@ def local_pallets_to_stock_via_pi?(explain: false)
...
+                      pre_stock: 'Pre-Stock Pallets',
+                      failed_stock: 'Failed Pallets',
+                      uninspected_stock: 'Uninspected Pallets' }
 
-      setting(:enable_depot_returns)
-    end
-
-    def can_allocate_pre_stock_pallets?(explain: false)
-      return 'Allow non stock pallets to be allocated to load.' if explain
+      return "Allowed allocation mode hash. #{valid_modes}" if explain
```
Comment (jcsjcs) [inline lib/client_rules/client_fg_rules.rb:436]: The explanation should only list the modes that are valid for the client.

File: lib/finished_goods/test/task_permission_checks/test_load.rb [hunk 4]
```diff
@@ -13,6 +13,7 @@ def entity(attrs = {})
       base_attrs = {
         id: 1,
         load_id: 1,
+        stock_allocation_mode: :standard,
```
Comment (jcsjcs) [inline lib/finished_goods/test/task_permission_checks/test_load.rb:16]: String, not symbol?

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: lib/client_rules/client_fg_rules.rb (hunk 2 above)
Comment (ViljoenFritz) [reply lib/client_rules/client_fg_rules.rb]: Added in table to get the values in the grid

**Pull request merged**